### PR TITLE
SIV support

### PR DESCRIPTION
**Pull request by reaperhulk**
Adds AES-SIV support to our AEAD module. Only supported in OpenSSL 3.0.0+

### CODE REVIEW

File: docs/hazmat/primitives/aead.rst [hunk 1]
```diff
@@ -228,6 +228,89 @@ also support providing integrity for associated data which is not encrypted.
...
+
+            SIV performs nonce-based authenticated encryption when a component of
+            the associated data is a nonce. The final associated data in the
+            list is used for the nonce.
+
+            Random nonces should have at least 128-bits of entropy. If a nonce is
+            reused with SIV authenticity is retained and confidentiality is only
+            compromised to the extent that an attacker can determine that the
+            same plaintext (and same associated data) was protected with the same
+            nonce and key.
```
Comment (alex) [inline docs/hazmat/primitives/aead.rst:282]: Maybe mention that if you don't provide a nonce encryption is entirely deterministic

File: src/cryptography/hazmat/backends/openssl/aead.py [hunk 2]
```diff
@@ -159,7 +183,14 @@ def _encrypt(
     backend.openssl_assert(res != 0)
     tag = backend._ffi.buffer(tag_buf)[:]
 
-    return processed_data + tag
+    if isinstance(cipher, AESSIV):
+        # RFC 5297 defines the output as IV || C, where the tag we generate is
+        # the "IV" and C is the ciphertext. This is the opposite of our
+        # other AEADs, which are Ciphertext || Tag
+        assert len(tag) == 16
```
Comment (alex) [inline src/cryptography/hazmat/backends/openssl/aead.py:190]: openssl asset

File: src/cryptography/hazmat/primitives/ciphers/aead.py [hunk 3]
```diff
@@ -289,3 +289,73 @@ def _check_params(
...
+        data: bytes,
+        associated_data: typing.Optional[typing.List[bytes]],
+    ) -> bytes:
+        if associated_data is None:
+            associated_data = []
+
+        self._check_params(data, associated_data)
+
+        if len(data) > self._MAX_SIZE or any(
+            [len(ad) > self._MAX_SIZE for ad in associated_data]
```
Comment (alex) [inline src/cryptography/hazmat/primitives/ciphers/aead.py:331]:
```suggestion
            len(ad) > self._MAX_SIZE for ad in associated_data
```

File: src/cryptography/hazmat/primitives/ciphers/algorithms.py [hunk 4]
```diff
@@ -28,7 +28,8 @@ class AES(CipherAlgorithm, BlockCipherAlgorithm):
     name = "AES"
     block_size = 128
     # 512 added to support AES-256-XTS, which uses 512-bit keys
-    key_sizes = frozenset([128, 192, 256, 512])
+    # 384 added to support AES-192-SIV (SIV uses split keys)
```
Comment (alex) [inline src/cryptography/hazmat/primitives/ciphers/algorithms.py:31]: Why is this needed

Comment (reaperhulk) [reply src/cryptography/hazmat/primitives/ciphers/algorithms.py:31]: ...It isn't and is a legacy of a separate approach.

**Review by alex — APPROVED** [review]
